Kong/kong
# RFC: Labels API in kong

Hello Kong Community, 
I'm proposing a feature to easily manage entities in Kong in groups, rather than configuring each one at atom(per api/consumer) level. This is a currently a prototype design and implementation.
**Comments/questions/feedback are welcome.**

## Motivation
- Grouping
  - While grouping entities is metadata and should be kept out of the core usually, however, in cases when it enhances capabilities of the core, it's good to include.
  - Any configuration change applied today is at atom level i.e. a user can manipulate an api or a consumer but not group of those simultaneously. This is not just a matter of inconvenience though. Providing filtering capabilities for plugins and other use-cases is not easy without replicating states.
- Replication of configurations
  - Consumers/apis share traits/properties in most use cases of API gateways. Every consumer is most probably not unique. It helps if some properties can be added/updated for a group, rather than iterating over all the entities.
  - Not having any aggregation feature, leads to a huge duplication here.
  - Example: There are 3 types of developers who use our APIs: Free, Paid, Enterprise. These give them different rate-limits. Currently, one would need to add a rate-limit plugin for every consumer, rather than having 3 configuration, we would have as many as consumers.
- Reporting type of consumer to upstream
  - Oauth scopes usually determine the permissions of a user(consumer in Kong terminology). 
  - While, this works very well, for apis not using Oauth, consumers could be classified/tagged by Kong since it makes sense to perform all auth related tasks at gateway level.

## Description
- Label/Tag/Group is a core entity; It has an id, unique name and could have a description as well.
- Most of the core entities can be associated with a label. Any entity can be associated with multiple labels.
- A label is also a common identification for a group of entity
- A plugin can be configured at label level as well. Example: key-auth plugin can be enabled for apis, which share a common label `authentication`
- A label can be used in place of consumers as well. A good example is ACL groups in the ACL plugin. One could white/black list consumers with specific label(s).

## Issues (in current design and implementation)
- Migration issue because of DAO changes; `2017-03-27-132300_anonymous` fails to execute because it can't find `label_id` column
- A plugin with a `label_id` is also a global plugin in application logic; No search interface for empty column from dao; we could add a boolean column `global` in the plugins table
- Data layer design
Ideally, keeping labels in apis table (same for consumes) as an array of foreign keys would work best; Q: What are other entities that could use labels? Upstreams? https://github.com/Kong/kong/commit/264b6503d3f155473f00141745bb2b07cd336ffd#diff-4ac32a78649ca5bdd8e0ba38b7006a1eL28 
- Could label's definition be even more broader?
- Tag instead of Label as the name of the entity?

## TODO
- Migrations
  - Add migrations for Cassandra
  - Indexes in migrations
  - Add index on labels(name)
  - Add conditional unique index on label_mappings
- Validations for label_mappings
  - Add self_check function to enforce not null for combinations: (label_id,consumer_id) and (label_id,api_id) at the application layer
- **Define Behavior** when multiple plugins of same kind can be executed in a request due to multiple labels; label A and label B, both have key-auth, now if an api is added to labels A and B, then what should be the behavior?
- Expose labels to upstream api, (configurable per api)?
- Check/verify REST design of /entity/:id/labels
- Add validations for POST /entity/:id/labels/ request
- Add the following rest endpoints, need to figure how to paginate
  - GET /labels/:label_name_or_id/apis 
  - GET /labels/:label_name_or_id/consumers
- Caching
  - Completely figure out how cache invalidation travels through in code;
  - Will cache invalidation work if cache_key does not contain the primary key?
  - Cache invalidations are currently not working; not sure
- Tests!
- Documentaiton
- Benchmark
  - What is the performance penalty?
